src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+        goto end;
+    }
+
+    /* Verify start value (from AN2013-004b). */
+    if (!DNP3CheckStartBytes(hdr)) {
+        SCLogDebug("Invalid start bytes.");
+        return ALPROTO_FAILED;
+    }
+
+    /* Verify minimum length. */
+    if (hdr->len < DNP3_MIN_LEN) {
+        SCLogDebug("Packet too small to be a valid DNP3 fragment.");
+        return ALPROTO_FAILED;
+    }
+
+end:
+    SCLogDebug("Detected DNP3.");
+    return ALPROTO_DNP3;
+}
+
+/**
+ * \brief Calculate a DNP3 session hash for the session hash table.
+ */
+static uint32_t DNP3SessionHash(HashListTable *ht, void *buf, uint16_t buflen)
+{
+    uint16_t a = 0, b = 0;
+    uint32_t hash;
+
+    if (buflen == sizeof(DNP3Session)) {
+        DNP3Session *session = buf;
+        a = session->master;
+        b = session->slave;
+    }
+    else if (buflen == sizeof(DNP3LinkHeader)) {
+        DNP3LinkHeader *header = buf;
+        a = header->src;
+        b = header->dst;
+    }
+
+    hash = (a * b) % ht->array_size;
+    return hash;
+}
+
+/**
+ * \brief Compare DNP3 session hash - for the session hash table.
+ */
+static char DNP3SessionHashCompare(void *a, uint16_t alen, void *b,
+    uint16_t blen)
+{
+    uint16_t a1 = 0, a2 = 0;
+    uint16_t b1 = 0, b2 = 0;
+
+    if (alen == sizeof(DNP3Session)) {
+        DNP3Session *session = a;
+        a1 = session->master;
+        a2 = session->slave;
+    }
+    else if (alen == sizeof(DNP3LinkHeader)) {
+        DNP3LinkHeader *header = a;
+        a1 = header->src;
+        a2 = header->dst;
+    }
+
+    if (blen == sizeof(DNP3Session)) {
+        DNP3Session *session = b;
+        b1 = session->master;
+        b2 = session->slave;
+    }
+    else if (blen == sizeof(DNP3LinkHeader)) {
+        DNP3LinkHeader *header = b;
+        b1 = header->src;
+        b2 = header->dst;
+    }
+
+    if (a1 == b1 && a2 == b2) {
+        return 1;
+    }
+    if (a1 == b2 && a2 == b1) {
+        return 1;
+    }
+
+    return 0;
+}
+
+/**
+ * \brief Reassemble the application layer by stripping the CRCs.
+ *
+ * \param input Input buffer starting at the transport header (which
+ *    will be removed from the output).
+ * \param input_len Length of the input buffer.
+ * \param output Pointer to output buffer (may be realloc'd).
+ * \param output_len Pointer to output length.
+ *
+ * \retval Pointer to output buffer if successful, otherwise NULL.
+ */
+static uint8_t *DNP3ReassembleApplicationLayer(uint8_t *input,
+    uint32_t input_len, uint8_t **output, uint32_t *output_len)
+{
+    int len = (input_len / (DNP3_BLOCK_SIZE + DNP3_CRC_LEN)) * DNP3_BLOCK_SIZE;
+    int rem = input_len % (DNP3_BLOCK_SIZE + DNP3_CRC_LEN);
+
+    if (rem) {
+        /* The remainder must be at least one byte plus a CRC. */
+        if (rem < 3) {
+            return NULL;
+        }
+        len += rem + DNP3_CRC_LEN;
+    }
+
+    /* Remove one byte for the transport header which won't be
+     * included in the output. */
+    len--;
```

Comment:
Length is now checked for 0 prior to subtraction, then checked for minimum valid length after subtraction.